In [1]:
# Import libraries
import pandas as pd
import numpy as np

# Question 1

What's the version of Pandas that you installed?

In [2]:
pd.__version__

'2.0.3'

# Getting the data

For this homework, we'll use the California Housing Prices dataset.

You can do it with wget:

In [3]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv

--2023-09-12 23:28:48--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv.2’

housing.csv.2       100%[===================>]   1.36M  4.62MB/s    in 0.3s    

2023-09-12 23:28:48 (4.62 MB/s) - ‘housing.csv.2’ saved [1423529/1423529]



In [4]:
!ls

Homework.ipynb	housing.csv  housing.csv.1  housing.csv.2


In [5]:
FILE_NAME = 'housing.csv'

In [6]:
df = pd.read_csv(FILE_NAME)
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

# Question 2

How many columns are in the dataset?
- 10
- 6560
- 10989
- 20640

In [7]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity'],
      dtype='object')

# Question 3

Which columns in the dataset have missing values?
- total_rooms
- total_bedrooms
- both of the above
- no empty columns in the dataset

In [8]:
df.isnull().any(axis=0)

longitude             False
latitude              False
housing_median_age    False
total_rooms           False
total_bedrooms         True
population            False
households            False
median_income         False
median_house_value    False
ocean_proximity       False
dtype: bool

#      Question 4

How many unique values does the ocean_proximity column have?
- 3
- 5
- 7
- 9

In [9]:
df.ocean_proximity.nunique()

5

# Question 5

What's the average value of the median_house_value for the houses located near the bay?
- 49433
- 124805
- 259212
- 380440

In [10]:
df[df.ocean_proximity=='NEAR BAY'].median_house_value.mean()

259212.31179039303

# Question 6

- Calculate the average of total_bedrooms column in the dataset.
- Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
- Now, calculate the average of total_bedrooms again.

Has it changed?

Has it changed?
- Yes
- No


In [11]:
print(df.total_bedrooms.mean())
df.total_bedrooms.fillna(df.total_bedrooms.mean())
print(df.total_bedrooms.fillna(df.total_bedrooms.mean()).mean())

537.8705525375618
537.8705525375617


# Question 7

1. Select all the options located on islands.
2. Select only columns housing_median_age, total_rooms, total_bedrooms.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
5. Compute the inverse of XTX.
6. Create an array y with values [950, 1300, 800, 1000, 1300].
7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
8. What's the value of the last element of w?

In [12]:
X = df[df.ocean_proximity=='ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']].to_numpy()
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [13]:
XTX = np.dot(X.T,X)
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [14]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [15]:
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [16]:
w= (XTX_inv@X.T)@y
w

array([23.12330961, -1.48124183,  5.69922946])